sql/lease.go

```diff
@@ -0,0 +1,500 @@
+		s := l.data[i]
+		if s.Version == version {
+			return s.expiration >= expiration
+		}
+		return s.Version > version
+	})
+	if i < len(l.data) {
+		s := l.data[i]
+		if s.Version == version && s.expiration == expiration {
+			return i, true
+		}
+	}
+	return i, false
+}
+
+func (l *leaseSet) findNewest(version uint32) *LeaseState {
+	if len(l.data) == 0 {
+		return nil
+	}
+	if version == 0 {
+		// No explicitly version, return the newest lease of the latest version.
+		return l.data[len(l.data)-1]
+	}
+	// Find the index of the first lease with version > targetVersion.
+	i := sort.Search(len(l.data), func(i int) bool {
+		return l.data[i].Version > version
+	})
+	if i == 0 {
+		return nil
+	}
+	// i-1 is the index of the newest lease for the previous version (the version
+	// we're looking for).
+	s := l.data[i-1]
+	if s.Version == version {
+		return s
+	}
+	return nil
+}
+
+type tableState struct {
+	id ID
+	mu sync.Mutex
+	// The active leases for the table: sorted by their version and expiration
+	// time. There may be more than one active lease when the system is
+	// transitioning from one version of the descriptor to another or when the
+	// node preemptively acquires a new lease for a version when the old lease
+	// has not yet expired.
+	active leaseSet
+	// A channel used to indicate whether a lease is actively being acquired.
+	// nil if there is no lease acquisition in progress for the table. If
+	// non-nil, the channel will be closed when lease acquisition completes.
+	acquiring chan struct{}
+}
+
+func (t *tableState) acquire(version uint32, store LeaseStore) (*LeaseState, error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	for {
+		s := t.active.findNewest(version)
+		if s != nil {
+			if version != 0 && s != t.active.findNewest(0) {
+				// If a lease was requested for an old version of the descriptor,
+				// return it even if there is only a short time left before it
+				// expires. We can't renew this lease as doing so would violate the
+				// invariant that we only get leases on the newest version. The
+				// transaction will either finish before the lease expires or it will
+				// abort, which is what will happen if we returned an error here.
+				s.refcount++
+				return s, nil
+			}
+			if s.expiration-store.clock.Now().WallTime >= minLeaseDuration {
+				s.refcount++
+				return s, nil
+			}
+		} else if version != 0 {
+			n := t.active.findNewest(0)
+			if n != nil && version < n.Version {
+				return nil, util.Errorf("table %d unable to acquire lease on old version: %d < %d",
+					t.id, version, n.Version)
+			}
+		}
+
+		if t.acquiring != nil {
+			// There is already a lease acquisition in progress. Wait for it to complete.
+			t.mu.Unlock()
+			<-t.acquiring
+			t.mu.Lock()
+		} else {
+			// There is no active lease acquisition so we'll go ahead and perform
+			// one.
+			t.acquiring = make(chan struct{})
+			t.mu.Unlock()
+			s, err := store.Acquire(t.id, version)
+			t.mu.Lock()
```

Comment:
I'm not fond of the unlock/lock that you do here and above. perhaps you want to move the code above the t.acquiring != nil into a separate function that holds a lock. Then we need some way to protect t.acquiring and t.active below.